service/src/test/kotlin/app/cash/backfila/service/BackfillRunnerTest.kt

```diff
@@ -589,6 +589,153 @@ class BackfillRunnerTest {
+        assertThat(getSinglePartitionCursor(runner)).isNull()
+
+        // The second request sends a partial result and the rest return normally
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+
+        waitForOtherCoroutines()
+        var firstBatchCursor = getSinglePartitionCursor(runner)
+        checkNotNull(firstBatchCursor) // cursor was set when the first batch succeeded
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().partial_batch_bookmark(bookmark).build()))
+
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is unchanged
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is unchanged
+
+        // Only two requests are sent. The first is in response to the first batch succeeding and
+        // the second is the follow up request
+        val queuedRequest = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        assertThat(queuedRequest).isNotNull
+        val followupRequest = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        assertThat(bookmarkRequest.batch_range).isEqualTo(followupRequest.batch_range)
+        assertThat(followupRequest.partial_batch_bookmark).isEqualTo(bookmark)
+
+        waitForOtherCoroutines()
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.poll()).isNull()
+
+        // complete the two stuck requests currently being processed
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is still unchanged
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+        waitForOtherCoroutines()
+        val cursor = getSinglePartitionCursor(runner)
+        assertThat(cursor).isGreaterThan(firstBatchCursor) // cursor increased
+        assertThat(cursor).isEqualTo(queuedRequest.batch_range.end.utf8().toLong())
+
+        // And we can get another 3
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+
+        waitForOtherCoroutines()
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.poll()).isNull()
+      } finally {
+        runner.stop()
+      }
+    }
+  }
+
+  private suspend fun waitForOtherCoroutines() {
```

Comment:
This is slightly better than having that delay everywhere. Maybe this is sufficient?

Reply:
seems fine